### PR TITLE
[WIP]nxos_facts requires JSON structured output support

### LINKED ISSUE #41514: nxos_facts: AttributeError: 'unicode' object has no attribute 'get'
##### SUMMARY

When running task with `nxos_facts`, getting AttributeError.

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME

nxos_facts

##### ANSIBLE VERSION

```
ansible 2.5.4
  config file = /Users/jvanderaa/Documents/ansible.cfg
  configured module search path = [u'/Users/jvanderaa/Documents/Ansible/library']
  ansible python module location = /Users/jvanderaa/Library/Python/2.7/lib/python/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.10 (default, Oct  6 2017, 22:29:07) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)]
```

##### OS / ENVIRONMENT

MacOS 10.13.4

##### STEPS TO REPRODUCE

```yaml
    - name: NXOS>> Get configuration
      nxos_facts:
        gather_subset: config
```
vars from hostfile
```ini
ansible_network_os=nxos
```

##### EXPECTED RESULTS

I'd expect to have the configuration gathered
##### ACTUAL RESULTS

```
The full traceback is:
Traceback (most recent call last):
  File "/var/folders/dq/_tjy3jfn0j98n6j590nyr24jpxb_7r/T/ansible_MZ9Rcf/ansible_module_nxos_facts.py", line 602, in <module>
    main()
  File "/var/folders/dq/_tjy3jfn0j98n6j590nyr24jpxb_7r/T/ansible_MZ9Rcf/ansible_module_nxos_facts.py", line 585, in main
    inst.populate()
  File "/var/folders/dq/_tjy3jfn0j98n6j590nyr24jpxb_7r/T/ansible_MZ9Rcf/ansible_module_nxos_facts.py", line 233, in populate
    if data.get('sys_ver_str'):
AttributeError: 'unicode' object has no attribute 'get'

fatal: [nxos01-1]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/var/folders/dq/_tjy3jfn0j98n6j590nyr24jpxb_7r/T/ansible_MZ9Rcf/ansible_module_nxos_facts.py\", line 602, in <module>\n    main()\n  File \"/var/folders/dq/_tjy3jfn0j98n6j590nyr24jpxb_7r/T/ansible_MZ9Rcf/ansible_module_nxos_facts.py\", line 585, in main\n    inst.populate()\n  File \"/var/folders/dq/_tjy3jfn0j98n6j590nyr24jpxb_7r/T/ansible_MZ9Rcf/ansible_module_nxos_facts.py\", line 233, in populate\n    if data.get('sys_ver_str'):\nAttributeError: 'unicode' object has no attribute 'get'\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 1
}
```

### LINKED ISSUE #41828: nxos_facts fails on gathering memtotal_mb fact
##### SUMMARY
Running the nxos_facts task against a Nexus7009 running 6.2(16) and 5596 5.2(1)N1(9) fails with the same error, unable to gather information about the memtotal_mb fact. 

This does work against a N3K 31108TCV version 7.0(3)I4(6)

Checking what I believe to be the specific output being parsed incorrectly (show system resources) then here are sample outputs for each.

Nexus 7K running 6.2(16) - Fails with error captured below
```
N7K# show system resources
Load average:   1 minute: 0.25   5 minutes: 0.14   15 minutes: 0.10
Processes   :   502 total, 2 running
CPU states  :   0.18% user,   1.23% kernel,   98.58% idle
        CPU0 states  :   0.99% user,   4.95% kernel,   94.05% idle
        CPU1 states  :   0.00% user,   0.94% kernel,   99.05% idle
        CPU2 states  :   0.98% user,   4.90% kernel,   94.11% idle
        CPU3 states  :   0.00% user,   0.00% kernel,   100.00% idle
        CPU4 states  :   0.00% user,   0.00% kernel,   100.00% idle
        CPU5 states  :   0.00% user,   3.80% kernel,   96.19% idle
        CPU6 states  :   0.00% user,   0.00% kernel,   100.00% idle
        CPU7 states  :   0.00% user,   0.00% kernel,   99.99% idle
        CPU8 states  :   0.91% user,   0.91% kernel,   98.16% idle
        CPU9 states  :   0.00% user,   3.44% kernel,   96.55% idle
        CPU10 states  :   0.00% user,   0.00% kernel,   99.99% idle
        CPU11 states  :   0.00% user,   0.81% kernel,   99.18% idle
        CPU12 states  :   0.00% user,   0.00% kernel,   100.00% idle
        CPU13 states  :   0.00% user,   0.00% kernel,   100.00% idle
        CPU14 states  :   0.00% user,   0.00% kernel,   100.00% idle
        CPU15 states  :   0.00% user,   0.00% kernel,   100.00% idle
Memory usage:   32745060K total,   3420196K used,   29324864K free
Current memory status: OK
```
N5K Running 5.2(1)N1(9) - Fails with error captured below
```
N5K# show system resources
Load average:   1 minute: 0.05   5 minutes: 0.08   15 minutes: 0.02
Processes   :   313 total, 2 running
CPU states  :   0.5% user,   0.0% kernel,   99.5% idle
Memory usage:   8263840K total,   1832424K used,   6431416K free
``` 
N3K Running 7.0(3)I4(6) - Success, module completes as normal
```
N3K# show system resources
Load average:   1 minute: 0.46   5 minutes: 0.44   15 minutes: 0.43
Processes   :   584 total, 4 running
CPU states  :   0.76% user,   3.81% kernel,   95.41% idle
        CPU0 states  :   0.00% user,   1.03% kernel,   98.96% idle
        CPU1 states  :   1.98% user,   2.97% kernel,   95.04% idle
        CPU2 states  :   0.00% user,   1.03% kernel,   98.96% idle
        CPU3 states  :   2.02% user,   10.10% kernel,   87.87% idle
Memory usage:   16401416K total,   4806884K used,   11594532K free
Current memory status: OK
```
So for whatever reason the "Memory usage" row isn't being parsed for the two failed outputs. Not catching offhand what the difference is.

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
nxos_facts

##### ANSIBLE VERSION

```
ansible 2.5.5
  config file = /mnt/c/Users/ataylor/Documents/ansible-working/ansible.cfg
  configured module search path = [u'/home/ataylor/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]
```

##### CONFIGURATION
PARAMIKO_HOST_KEY_AUTO_ADD(/ansible-working/ansible.cfg) = True

##### OS / ENVIRONMENT

Host OS: N/A
Target hosts: Nexus 7K NX-OS 6.2(16) , Nexus 5K 5.2(1)N1(9)

##### STEPS TO REPRODUCE
Run the nxos_facts module against the specified software versions, and output fails.

```yaml
 - name: Get facts (NXOS)
   nxos_facts:
    gather_subset: all
   when: ansible_network_os == "nxos"
```

##### EXPECTED RESULTS
Would expect the facts to be populated as it does for the N3K running version 7.

##### ACTUAL RESULTS
```
<192.168.1.1> attempting to start connection
<192.168.1.1> using connection plugin network_cli
<192.168.1.1> found existing local domain socket, using it!
<192.168.1.1> updating play_context for connection
<192.168.1.1>
<192.168.1.1> local domain socket path is /.ansible/pc/d894e2db50
Using module file /usr/lib/python2.7/dist-packages/ansible/modules/network/nxos/nxos_facts.py
<192.168.1.1> ESTABLISH LOCAL CONNECTION FOR USER: ataylor
<192.168.1.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /.ansible/tmp/ansible-local-1557eZYAor/ansible-tmp-1529632472.4-112453209332856 `" && echo ansible-tmp-1529632472.4-112453209332856="` echo /.ansible/tmp/ansible-local-1557eZYAor/ansible-tmp-1529632472.4-112453209332856 `" ) && sleep 0'
<192.168.1.1> PUT /.ansible/tmp/ansible-local-1557eZYAor/tmpDQvtIM TO /.ansible/tmp/ansible-local-1557eZYAor/ansible-tmp-1529632472.4-112453209332856/nxos_facts.py
<192.168.1.1> EXEC /bin/sh -c 'chmod u+x /.ansible/tmp/ansible-local-1557eZYAor/ansible-tmp-1529632472.4-112453209332856/ /.ansible/tmp/ansible-local-1557eZYAor/ansible-tmp-1529632472.4-112453209332856/nxos_facts.py && sleep 0'
<192.168.1.1> EXEC /bin/sh -c '/usr/bin/python /.ansible/tmp/ansible-local-1557eZYAor/ansible-tmp-1529632472.4-112453209332856/nxos_facts.py && sleep 0'
<192.168.1.1> EXEC /bin/sh -c 'rm -f -r /.ansible/tmp/ansible-local-1557eZYAor/ansible-tmp-1529632472.4-112453209332856/ > /dev/null 2>&1 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_dt6HU3/ansible_module_nxos_facts.py", line 602, in <module>
    main()
  File "/tmp/ansible_dt6HU3/ansible_module_nxos_facts.py", line 585, in main
    inst.populate()
  File "/tmp/ansible_dt6HU3/ansible_module_nxos_facts.py", line 255, in populate
    self.facts['memtotal_mb'] = int(data['memory_usage_total']) / 1024
TypeError: string indices must be integers

fatal: [192.168.1.1]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_dt6HU3/ansible_module_nxos_facts.py\", line 602, in <module>\n    main()\n  File \"/tmp/ansible_dt6HU3/ansible_module_nxos_facts.py\", line 585, in main\n    inst.populate()\n  File \"/tmp/ansible_dt6HU3/ansible_module_nxos_facts.py\", line 255, in populate\n    self.facts['memtotal_mb'] = int(data['memory_usage_total']) / 1024\nTypeError: string indices must be integers\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 1
```

### PR DESCRIPTION
**Pull request by trishnaguha**
Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

##### SUMMARY

fixes https://github.com/ansible/ansible/issues/41828
fixes https://github.com/ansible/ansible/issues/41514
Add notes that JSON structured output is required on NX-OS device to use this module.

##### ISSUE TYPE

 - Bugfix Pull Request
 - Docs Pull Request

##### COMPONENT NAME

modules/network/nxos/nxos_facts.py
##### ANSIBLE VERSION

```
devel, 2.6
```

### CODE REVIEW

**Comment by jvanderaa** [comment]
So what would be used instead of NXOS module to use Ansible against devices that don't support json?

File: lib/ansible/modules/network/nxos/nxos_facts.py [hunk 1]
```diff
@@ -546,6 +550,13 @@ def main():
     runable_subsets = set()
     exclude_subsets = set()
 
+    capabilities = get_capabilities(module)
+    if capabilities:
+        os_version = capabilities['device_info']['network_os_version']
+        os_version_major = int(os_version[0])
+        if os_version_major < 7:
```
Comment (mikewiebe) [inline lib/ansible/modules/network/nxos/nxos_facts.py:557]: @trishnaguha We need to support the `nxos_facts` module on the following versions.

1. 7.x and later
2. 6.0(2)A8(1) and later

You could probably just check for (> 7) and `6.0(2)A8`

Comment (trishnaguha) [reply lib/ansible/modules/network/nxos/nxos_facts.py:557]: @mikewiebe Thanks. This PR is moved to https://github.com/ansible/ansible/pull/42089. Please review https://github.com/ansible/ansible/pull/42089/files#diff-83a6bf7294d0c5ac59ee458de130f795R549 and let me know if that covers all the cases.

**Comment by jvanderaa** [comment]
I don't believe that this is the proper failed scenario. The device referenced in this case is running a code version > 7, 

```cisco
system:    version 7.0(8)N1(1)
```

**Comment by trishnaguha** [comment]
@jvanderaa what is the platform you are using?

**Comment by jvanderaa** [comment]
```cisco
"Nexus5548 Chassis"
```

**Comment by trishnaguha** [comment]
@jvanderaa Addressed in PR https://github.com/ansible/ansible/pull/42729

**Pull request closed without merging**